### PR TITLE
Private-in-public lints implementation

### PR DESCRIPTION
**Pull request by Bryanskiy**
Next part of RFC https://github.com/rust-lang/rust/issues/48054. 

r? @petrochenkov

### CODE REVIEW

File: compiler/rustc_privacy/src/lib.rs [hunk 1]
```diff
@@ -409,8 +405,24 @@ impl VisibilityLike for ty::Visibility {
     }
 }
 
-impl VisibilityLike for Option<EffectiveVisibility> {
-    const MAX: Self = Some(EffectiveVisibility::from_vis(ty::Visibility::Public));
+struct FindMinEffectiveVis(EffectiveVisibility);
```
Comment (petrochenkov) [inline compiler/rustc_privacy/src/lib.rs:408]: Could you actually make two wrapper structures
```rust
struct ShallowEffectiveVis(EffectiveVisibility);
struct NonShallowEffectiveVis(EffectiveVisibility);
```
so it's immediately clear what's the difference?

File: compiler/rustc_resolve/src/effective_visibilities.rs [hunk 2]
```diff
@@ -81,7 +81,7 @@ impl<'r, 'a, 'tcx> EffectiveVisibilitiesVisitor<'r, 'a, 'tcx> {
             def_effective_visibilities: Default::default(),
             import_effective_visibilities: Default::default(),
             current_private_vis: Visibility::Restricted(CRATE_DEF_ID),
-            changed: false,
+            changed: true,
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/effective_visibilities.rs:84]: Could you split this into a separate commit (with a test)?
This is a bugfix unrelated to private-in-public lints.

File: library/std/src/sys/unix/fd.rs [hunk 3]
```diff
@@ -425,6 +427,7 @@ impl FileDesc {
         }
     }
     #[cfg(any(target_os = "espidf", target_os = "horizon"))]
+    #[allow(dead_code)]
```
Comment (petrochenkov) [inline library/std/src/sys/unix/fd.rs:430]: Is this still necessary?
Did you figure out why exactly is it reported as dead code now?

File: compiler/rustc_privacy/src/lib.rs [hunk 4]
```diff
@@ -422,13 +434,15 @@ impl VisibilityLike for Option<EffectiveVisibility> {
...
     const SHALLOW: bool = true;
     fn new_min(find: &FindMin<'_, '_, Self>, def_id: LocalDefId) -> Self {
-        if let Some(min) = find.min {
-            return find
-                .effective_visibilities
-                .effective_vis(def_id)
-                .map(|eff_vis| min.min(*eff_vis, find.tcx));
-        }
-        None
+        let private_vis = ty::Visibility::Restricted(find.tcx.parent_module_from_def_id(def_id));
```
Comment (petrochenkov) [inline compiler/rustc_privacy/src/lib.rs:437]: This is likely a perf regression, I'd rather keep the `Option`.

File: compiler/rustc_privacy/src/lib.rs [hunk 5]
```diff
@@ -1749,12 +1759,15 @@ struct SearchInterfaceForPrivateItemsVisitor<'tcx> {
     item_def_id: LocalDefId,
     /// The visitor checks that each component type is at least this visible.
     required_visibility: ty::Visibility,
+    effective_vis: Option<EffectiveVisibility>,
```
Comment (petrochenkov) [inline compiler/rustc_privacy/src/lib.rs:1762]:
```suggestion
    required_vis: ty::Visibility,
    required_effective_vis: Option<EffectiveVisibility>,
```
for consistency.

File: compiler/rustc_privacy/src/lib.rs [hunk 6]
```diff
@@ -1862,6 +1878,29 @@ impl SearchInterfaceForPrivateItemsVisitor<'_> {
             }
         }
 
+        let Some(effective_vis) = self.effective_vis else {
+            return false;
+        };
+        let reexported_ev = *effective_vis.at_level(Level::Reexported);
```
Comment (petrochenkov) [inline compiler/rustc_privacy/src/lib.rs:1884]:
```suggestion
        let reachable_at_vis = *effective_vis.at_level(Level::Reachable);
```
That's what the RFC says at least.
Were there any issues with this formulation?

File: compiler/rustc_privacy/src/lib.rs [hunk 7]
```diff
@@ -1862,6 +1878,29 @@ impl SearchInterfaceForPrivateItemsVisitor<'_> {
...
+
+        if !vis.is_at_least(reexported_ev, self.tcx) {
+            if self.in_primary_interface {
```
Comment (petrochenkov) [inline compiler/rustc_privacy/src/lib.rs:1887]:
```suggestion
            let lint = if self.in_primary_interface { lint::builtin::PRIVATE_INTERFACE } else { lint::builtin::PRIVATE_BOUNDS };
```
Everything else is common.

File: compiler/rustc_privacy/src/lib.rs [hunk 8]
```diff
@@ -1895,25 +1934,49 @@ impl<'tcx> DefIdVisitor<'tcx> for SearchInterfaceForPrivateItemsVisitor<'tcx> {
...
+            in_primary_interface: true,
+        }
+    }
+
+    fn check_unnameable(&self, def_id: LocalDefId, effective_vis: Option<EffectiveVisibility>) {
+        let Some(effective_vis) = effective_vis else {
+            return;
+        };
+
+        let reexported_ev = effective_vis.at_level(Level::Reexported);
```
Comment (petrochenkov) [inline compiler/rustc_privacy/src/lib.rs:1968]:
```suggestion
        let reexported_at_vis = effective_vis.at_level(Level::Reexported);
```
Nit: this is `Visibility`, not `EffectiveVisibility`.

File: compiler/rustc_privacy/src/lib.rs [hunk 9]
```diff
@@ -1895,25 +1934,49 @@ impl<'tcx> DefIdVisitor<'tcx> for SearchInterfaceForPrivateItemsVisitor<'tcx> {
...
+        let reachable_ev = effective_vis.at_level(Level::Reachable);
+
+        if !reexported_ev.is_at_least(*reachable_ev, self.tcx) {
```
Comment (petrochenkov) [inline compiler/rustc_privacy/src/lib.rs:1971]:
```suggestion
        if reexported_ev != reachable_ev {
```
reachable_vis is always larger or equal than reexported_vis, so equality is the only case in which this condition can hold.

File: compiler/rustc_privacy/src/lib.rs [hunk 10]
```diff
@@ -1922,8 +1985,10 @@ impl<'tcx> PrivateItemsInPublicInterfacesChecker<'tcx> {
         def_id: LocalDefId,
         assoc_item_kind: AssocItemKind,
         vis: ty::Visibility,
+        effective_vis: Option<EffectiveVisibility>,
     ) {
-        let mut check = self.check(def_id, vis);
+        self.check_unnameable(def_id, self.get(def_id));
```
Comment (petrochenkov) [inline compiler/rustc_privacy/src/lib.rs:1990]:
```suggestion
        if is_assoc_ty { self.check_unnameable(def_id, self.get(def_id)); }
```
The `unnameable_types` lint should probably only be reported for *types* (or traits).

File: compiler/rustc_privacy/src/lib.rs [hunk 11]
```diff
@@ -1936,50 +2001,72 @@ impl<'tcx> PrivateItemsInPublicInterfacesChecker<'tcx> {
...
         let tcx = self.tcx;
         let def_id = id.owner_id.def_id;
         let item_visibility = tcx.local_visibility(def_id);
+        let effective_vis = self.get(def_id);
         let def_kind = tcx.def_kind(def_id);
 
         match def_kind {
             DefKind::Const | DefKind::Static(_) | DefKind::Fn | DefKind::TyAlias => {
-                self.check(def_id, item_visibility).generics().predicates().ty();
+                self.check_unnameable(def_id, effective_vis);
```
Comment (petrochenkov) [inline compiler/rustc_privacy/src/lib.rs:2017]:
```suggestion
                if DefKind::TyAlias { self.check_unnameable(def_id, effective_vis); }
```

File: compiler/rustc_privacy/src/lib.rs [hunk 12]
```diff
@@ -1936,50 +2001,72 @@ impl<'tcx> PrivateItemsInPublicInterfacesChecker<'tcx> {
...
                 if let hir::ItemKind::Enum(ref def, _) = item.kind {
-                    self.check(item.owner_id.def_id, item_visibility).generics().predicates();
+                    self.check_unnameable(item.owner_id.def_id, effective_vis);
+
+                    self.check(item.owner_id.def_id, item_visibility, effective_vis)
+                        .generics()
+                        .predicates();
 
                     for variant in def.variants {
+                        self.check_unnameable(variant.def_id, self.get(variant.def_id));
```
Comment (petrochenkov) [inline compiler/rustc_privacy/src/lib.rs:2066]: This is not exactly a type, but there's an RFC about making them types, so let's keep this for now.

File: compiler/rustc_privacy/src/lib.rs [hunk 13]
```diff
@@ -1936,50 +2001,72 @@ impl<'tcx> PrivateItemsInPublicInterfacesChecker<'tcx> {
...
                         for field in variant.data.fields() {
-                            self.check(field.def_id, item_visibility).ty();
+                            self.check_unnameable(field.def_id, self.get(field.def_id));
```
Comment (petrochenkov) [inline compiler/rustc_privacy/src/lib.rs:2068]: Not a type, can be removed.

File: compiler/rustc_privacy/src/lib.rs [hunk 14]
```diff
@@ -1989,8 +2076,11 @@ impl<'tcx> PrivateItemsInPublicInterfacesChecker<'tcx> {
                 let item = tcx.hir().item(id);
                 if let hir::ItemKind::ForeignMod { items, .. } = item.kind {
                     for foreign_item in items {
+                        let ev = self.get(foreign_item.id.owner_id.def_id);
                         let vis = tcx.local_visibility(foreign_item.id.owner_id.def_id);
-                        self.check(foreign_item.id.owner_id.def_id, vis)
+
+                        self.check_unnameable(foreign_item.id.owner_id.def_id, ev);
```
Comment (petrochenkov) [inline compiler/rustc_privacy/src/lib.rs:2082]:
```suggestion
                        if ForeignItemKind::Type { self.check_unnameable(foreign_item.id.owner_id.def_id, ev); }
```

File: compiler/rustc_privacy/src/lib.rs [hunk 15]
```diff
@@ -2003,11 +2093,22 @@ impl<'tcx> PrivateItemsInPublicInterfacesChecker<'tcx> {
...
+                    self.check(item.owner_id.def_id, item_visibility, effective_vis)
+                        .generics()
+                        .predicates();
 
                     for field in struct_def.fields() {
                         let field_visibility = tcx.local_visibility(field.def_id);
-                        self.check(field.def_id, min(item_visibility, field_visibility, tcx)).ty();
+                        let field_ev = self.get(field.def_id);
+
+                        self.check_unnameable(field.def_id, field_ev);
```
Comment (petrochenkov) [inline compiler/rustc_privacy/src/lib.rs:2105]: Not a type, can be removed.

File: compiler/rustc_privacy/src/lib.rs [hunk 16]
```diff
@@ -2020,10 +2121,20 @@ impl<'tcx> PrivateItemsInPublicInterfacesChecker<'tcx> {
                 if let hir::ItemKind::Impl(ref impl_) = item.kind {
                     let impl_vis =
                         ty::Visibility::of_impl(item.owner_id.def_id, tcx, &Default::default());
+                    let impl_ev = Some(
```
Comment (petrochenkov) [inline compiler/rustc_privacy/src/lib.rs:2124]: Need a comment here explaining that we are using the non-shallow version here, unlike when building the eff vis table, because even if `Pub<Priv>` *can* leak things in general, it rarely happens in practice, and here we need a low number of false positives.

File: compiler/rustc_privacy/src/lib.rs [hunk 17]
```diff
@@ -2020,10 +2121,20 @@ impl<'tcx> PrivateItemsInPublicInterfacesChecker<'tcx> {
...
+                        FindMinEffectiveVis::of_impl(
+                            item.owner_id.def_id,
+                            tcx,
+                            self.effective_visibilities,
+                        )
+                        .0,
+                    );
+
+                    self.check_unnameable(item.owner_id.def_id, effective_vis);
```
Comment (petrochenkov) [inline compiler/rustc_privacy/src/lib.rs:2133]: Not a type, can be removed.

**Comment by Bryanskiy** [comment]
Blocked on https://github.com/rust-lang/rust/pull/112426

File: compiler/rustc_resolve/src/effective_visibilities.rs (hunk 2 above)
Comment (Bryanskiy) [reply compiler/rustc_resolve/src/effective_visibilities.rs:84]: Done: https://github.com/rust-lang/rust/pull/112426

File: compiler/rustc_privacy/src/lib.rs (hunk 4 above)
Comment (Bryanskiy) [reply compiler/rustc_privacy/src/lib.rs:441]: These are necessary changes. Imp's should always be in the table. Here is an example: https://github.com/rust-lang/rust/pull/112426

File: library/std/src/sys/unix/fd.rs (hunk 3 above)
Comment (Bryanskiy) [reply library/std/src/sys/unix/fd.rs]: This is not necessary, it was reported due to misunderstanding with references.

File: compiler/rustc_lint_defs/src/builtin.rs [hunk 18]
```diff
@@ -4251,3 +4254,90 @@ declare_lint! {
...
+    ///     impl crate::SemiPriv {
+    ///         pub fn f(_: Priv) {}
+    ///     }
+    /// }
+    ///
+    /// # fn main() {}
+    /// ```
+    ///
+    /// {{produces}}
+    pub PRIVATE_INTERFACES,
```
Comment (petrochenkov) [inline compiler/rustc_lint_defs/src/builtin.rs:4282]: The
```rust
    /// Having something private in primary interface guarantees that
    /// the item will be unusable from outer modules due to type privacy.
```
part of `### Explanation` was reasonable, no need to remove it.

File: tests/ui/privacy/where-priv-type.stderr [hunk 19]
```diff
@@ -8,59 +8,137 @@ LL | pub struct S
...
    = note: `#[warn(private_in_public)]` on by default
 
+warning: type `PrivTy` is more private than the item `S`
+  --> $DIR/where-priv-type.rs:25:1
+   |
+LL | struct PrivTy;
+   | ------------- but type `PrivTy` is only usable at visibility `pub(crate)`
+...
+LL | pub struct S
+   | ^^^^^^^^^^^^ struct `S` is reachable at visibility `pub`
```
Comment (petrochenkov) [inline tests/ui/privacy/where-priv-type.stderr:18]: Same issue with the order of labels as in https://github.com/rust-lang/rust/pull/112413.
Let's convert these labels to notes for now.

File: compiler/rustc_lint_defs/src/builtin.rs [hunk 20]
```diff
@@ -4251,3 +4254,95 @@ declare_lint! {
...
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// It is often expected that if you can obtain an object of type `T`, then
+    /// you can name the type `T` as well, this lint attempts to enforce this rule.
+    pub UNNAMEABLE_TYPES,
+    Allow,
+    "effective visibility is larger than the area in which they can be named"
```
Comment (petrochenkov) [inline compiler/rustc_lint_defs/src/builtin.rs:4347]:
```suggestion
    "effective visibility of a type is larger than the area in which it can be named"
```

File: compiler/rustc_privacy/src/errors.rs [hunk 21]
```diff
@@ -98,3 +98,32 @@ pub struct PrivateInPublicLint<'a> {
...
+// They will replace private-in-public errors and compatibility lints in future.
+// See https://rust-lang.github.io/rfcs/2145-type-privacy.html for more details.
+#[derive(LintDiagnostic)]
+#[diag(privacy_private_interface_or_bounds_lint)]
+pub struct PrivateInterfacesOrBoundsLint<'a> {
+    #[label(privacy_i_label)]
+    pub i_span: Span,
+    pub i_kind: &'a str,
+    pub i_descr: DiagnosticArgFromDisplay<'a>,
+    pub i_vis_descr: &'a str,
```
Comment (petrochenkov) [inline compiler/rustc_privacy/src/errors.rs:123]:
```suggestion
    pub item_vis_descr: &'a str,
```
Could you expand `i` to `item` everywhere in this structure?

File: compiler/rustc_privacy/src/lib.rs [hunk 22]
```diff
@@ -194,6 +194,16 @@ where
...
 
+struct LazyDefPathStr<'tcx> {
+    def_id: DefId,
+    tcx: TyCtxt<'tcx>,
+}
+impl<'tcx> fmt::Display for LazyDefPathStr<'tcx> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.tcx.def_path_str(self.def_id))
+    }
+}
```
Comment (petrochenkov) [inline compiler/rustc_privacy/src/lib.rs:205]: Nit: could you move this above `trait DefIdVisitor`? (And also format with an empty line between items.)
Right now it's copypasted right between two impls of `DefIdVisitorSkeleton`.

File: compiler/rustc_privacy/src/lib.rs [hunk 23]
```diff
@@ -422,13 +439,15 @@ impl VisibilityLike for Option<EffectiveVisibility> {
...
-                .map(|eff_vis| min.min(*eff_vis, find.tcx));
-        }
-        None
+        let private_vis = ty::Visibility::Restricted(find.tcx.parent_module_from_def_id(def_id));
+
+        let effective_vis = find
+            .effective_visibilities
+            .effective_vis(def_id)
+            .cloned()
+            .unwrap_or(EffectiveVisibility::from_vis(private_vis));
```
Comment (petrochenkov) [inline compiler/rustc_privacy/src/lib.rs:448]:
```suggestion
            .unwrap_or_else(|| EffectiveVisibility::from_vis(ty::Visibility::Restricted(find.tcx.parent_module_from_def_id(def_id))));
```
No need to calculate the private visibility eagerly.

File: compiler/rustc_privacy/src/lib.rs [hunk 24]
```diff
@@ -1862,6 +1889,40 @@ impl SearchInterfaceForPrivateItemsVisitor<'_> {
             }
         }
 
+        let Some(effective_vis) = self.required_effective_vis else {
+            return false;
+        };
+
+        // FIXME: `Level::Reachable` should be taken instead of `Level::Reexported`
```
Comment (petrochenkov) [inline compiler/rustc_privacy/src/lib.rs:1896]: Need some explanation about why it's not done now.

File: compiler/rustc_privacy/src/lib.rs [hunk 25]
```diff
@@ -2020,10 +2158,29 @@ impl<'tcx> PrivateItemsInPublicInterfacesChecker<'tcx> {
                 if let hir::ItemKind::Impl(ref impl_) = item.kind {
                     let impl_vis =
                         ty::Visibility::of_impl(item.owner_id.def_id, tcx, &Default::default());
+
+                    // we are using the non-shallow version here, unlike when building the
+                    // effective visisibilities table to avoid false positives. For example:
```
Comment (petrochenkov) [inline compiler/rustc_privacy/src/lib.rs:2163]:
```suggestion
                    // effective visisibilities table to avoid large number of false positives. For example:
```

File: tests/ui/privacy/unnameable_types.rs [hunk 26]
```diff
@@ -0,0 +1,30 @@
+#![allow(unused)]
+#![allow(private_in_public)]
+#![deny(unnameable_types)]
+
+mod m {
+    pub struct PubStruct(pub i32); //~ ERROR struct `PubStruct` is reachable but cannot be named
+
+    pub enum PubE { //~ ERROR enum `PubE` is reachable but cannot be named
+        V(i32), //~ ERROR variant `PubE::V` is reachable but cannot be named
```
Comment (petrochenkov) [inline tests/ui/privacy/unnameable_types.rs:9]: Let's drop this diagnostic for enum variants after all.
If it's almost always duplicated with the same diagnostic for enums it doesn't look good.

**Comment by petrochenkov** [comment]
r=me after extending the comment in https://github.com/rust-lang/rust/pull/111801#discussion_r1225236877 and squashing commits.

**Comment by pnkfelix** [comment]
* primary regressions were to stm32f4-0.14.0 check+opt incr, but by a really small amount (0.18%).
* marking as triaged.

@rustbot label: +perf-regression-triaged
